File: lab3/gr8sub1/DMITRIEVA_POLINA/src/thread_race.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <mutex>
#include <atomic>
#include <chrono>
#include <string>
#include <cstring>

typedef enum { MODE_UNSYNC, MODE_MUTEX, MODE_ATOMIC } sync_mode_t;

static long long shared_counter_unsync = 0;
static long long shared_counter_mutex = 0;
static std::atomic<long long> shared_counter_atomic(0);
static std::mutex counter_mutex;

struct thread_args_t {
    int thread_index;
    long long iterations_per_thread;
};

static inline long long now_monotonic_ms() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::steady_clock::now().time_since_epoch()
    ).count();
}

static void worker_unsync(thread_args_t* arg) {
    for (long long i = 0; i < arg->iterations_per_thread; i++) {
        shared_counter_unsync++;
    }
}

static void worker_mutex(thread_args_t* arg) {
    for (long long i = 0; i < arg->iterations_per_thread; i++) {
        std::lock_guard<std::mutex> lock(counter_mutex);
        shared_counter_mutex++;
    }
}

static void worker_atomic(thread_args_t* arg) {
    for (long long i = 0; i < arg->iterations_per_thread; i++) {
        shared_counter_atomic.fetch_add(1, std::memory_order_relaxed);
    }
}

static sync_mode_t parse_mode(const char* s) {
    if (strcmp(s, "unsync") == 0) return MODE_UNSYNC;
    if (strcmp(s, "mutex") == 0) return MODE_MUTEX;
    if (strcmp(s, "atomic") == 0) return MODE_ATOMIC;
    std::cerr << "Unknown mode: " << s << " (use: unsync|mutex|atomic)" << std::endl;
    exit(2);
}

int main(int argc, char** argv) {
    if (argc < 4) {
        std::cerr << "Usage: " << argv[0] << " <num_threads> <iterations_per_thread> <unsync|mutex|atomic>" << std::endl;
        return 1;
    }

    int num_threads = std::stoi(argv[1]);
    long long iters = std::stoll(argv[2]);
    sync_mode_t mode = parse_mode(argv[3]);

    if (num_threads <= 0 || iters < 0) {
        std::cerr << "Invalid arguments" << std::endl;
        return 1;
    }

    std::cout << "[thread_race] Starting with " << num_threads << " threads, " 
              << iters << " iterations per thread, mode: " << argv[3] << std::endl;

    shared_counter_unsync = 0;
    shared_counter_mutex = 0;
    shared_counter_atomic.store(0);

    std::vector<std::thread> threads;
    std::vector<thread_args_t> args(num_threads);

    long long start_ms = now_monotonic_ms();

    for (int i = 0; i < num_threads; i++) {
        args[i].thread_index = i;
        args[i].iterations_per_thread = iters;
        
        switch (mode) {
            case MODE_UNSYNC:
                threads.emplace_back(worker_unsync, &args[i]);
                break;
            case MODE_MUTEX:
                threads.emplace_back(worker_mutex, &args[i]);
                break;
            case MODE_ATOMIC:
                threads.emplace_back(worker_atomic, &args[i]);
                break;
        }
    }

    for (auto& t : threads) {
        t.join();
    }

    long long end_ms = now_monotonic_ms();
    long long expected = (long long)num_threads * iters;
    long long actual = 0;

    switch (mode) {
        case MODE_UNSYNC:
            actual = shared_counter_unsync;
            break;
        case MODE_MUTEX:
            actual = shared_counter_mutex;
            break;
        case MODE_ATOMIC:
            actual = shared_counter_atomic.load();
            break;
    }

    std::cout << "RESULTS: mode=" << argv[3] 
              << " threads=" << num_threads 
              << " iters_per_thread=" << iters 
              << " expected=" << expected 
              << " actual=" << actual 
              << " time_ms=" << (end_ms - start_ms) 
              << std::endl;

    return 0;
}

File: lab3/grp9sub1/PaniavinRe/src/thread_race.py
#!/usr/bin/env python3
import threading
import time
import sys
from threading import Lock

class Counter:
    def __init__(self):
        self.unsync_value = 0
        self.mutex_value = 0
        self.atomic_value = 0
        self.lock = Lock()
    
    def unsync_increment(self):
        self.unsync_value += 1
    
    def mutex_increment(self):
        with self.lock:
            self.mutex_value += 1
    
    def atomic_increment(self):
        # В Python обычные операции уже атомарны для int
        self.atomic_value += 1

def worker_unsync(counter, iterations):
    for _ in range(iterations):
        counter.unsync_increment()

def worker_mutex(counter, iterations):
    for _ in range(iterations):
        counter.mutex_increment()

def worker_atomic(counter, iterations):
    for _ in range(iterations):
        counter.atomic_increment()

def run_test(num_threads, iterations_per_thread, mode):
    counter = Counter()
    threads = []
    
    start_time = time.time()
    
    for _ in range(num_threads):
        if mode == "unsync":
            t = threading.Thread(target=worker_unsync, args=(counter, iterations_per_thread))
        elif mode == "mutex":
            t = threading.Thread(target=worker_mutex, args=(counter, iterations_per_thread))
        elif mode == "atomic":
            t = threading.Thread(target=worker_atomic, args=(counter, iterations_per_thread))
        else:
            print(f"Unknown mode: {mode}")
            return
        
        threads.append(t)
        t.start()
    
    for t in threads:
        t.join()
    
    end_time = time.time()
    elapsed_ms = int((end_time - start_time) * 1000)
    
    expected = num_threads * iterations_per_thread
    
    if mode == "unsync":
        actual = counter.unsync_value
    elif mode == "mutex":
        actual = counter.mutex_value
    else:
        actual = counter.atomic_value
    
    print(f"mode={mode} threads={num_threads} iters_per_thread={iterations_per_thread} expected={expected} actual={actual} time_ms={elapsed_ms}")

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python3 thread_race.py <num_threads> <iterations_per_thread> <unsync|mutex|atomic>")
        sys.exit(1)
    
    num_threads = int(sys.argv[1])
    iterations = int(sys.argv[2])
    mode = sys.argv[3]
    
    run_test(num_threads, iterations, mode)
